Spark Jupyter getting started docker compose

kevinjqliu opened the pull request:
# Description

This PR moves the `docker-compose-jupyter.yml` file (and the `notebooks/` directory), formerly in the top-level directory, into the `getting-started/spark/` folder. 

The purpose is to unify the "getting started" guides into the same directory. 

Fixes #110

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Documentation update
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

```
docker-compose -f getting-started/spark/docker-compose.yml up
```
Open the `SparkPolaris.ipynb` Jupyter notebook
Grab the `root principal credentials` from the Polaris service and replace in the notebook cell.
Run all cells in notebook

# Checklist:

Please delete options that are not relevant.

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] If adding new functionality, I have discussed my implementation with the community using the linked GitHub issue

kevinjqliu commented:
I want to make sure this is something we want to do before proceeding to add more to the PR

cc @collado-mike / @flyrain

flyrain commented:
Make sense to me. Thanks @kevinjqliu! Do we have any doc for its usage? We may add doc if not.

kevinjqliu commented:
@flyrain yep i'll have a README in here, similar to the trino one

flyrain commented:
Sounds good. We will need these doc to be in the Polaris doc site, like this https://polaris.apache.org/docs/overview/. I couldn't find Trino's doc there, this may involve doc publish and link. cc @jbonofre

kevinjqliu commented:
I see, this is the [README for trino](https://github.com/apache/polaris/blob/main/getting-started/trino/README.md). I'll add a similar README for spark. 

As a follow-up, we can change the Polaris doc to refer to these guides https://polaris.apache.org/docs/quickstart

collado-mike commented:
This looks good to me. We should change the name of the compose file to just `docker-compose.yml` so we don't have to specify the filename in the command line :)

kevinjqliu commented:
@collado-mike makes sense, will do. 

I have a question [on slack](https://apache-iceberg.slack.com/archives/C07HADW86HZ/p1726508953511979) about unable to assume the role `arn:aws:iam::631484165566:role/datalake-storage-integration-role` in the notebook, do you mind taking a look?

sfc-gh-emaynard commented on getting-started/spark/create-polaris-catalog.sh:
I wonder if it might be easier to use the CLI here

kevinjqliu replied:
might be, if you want a spark-shell. I think the jupyter notebook does a good job of explaining a lot of the concepts

sfc-gh-emaynard replied:
Sorry, I meant the `polaris` CLI instead of using `curl`

kevinjqliu replied:
ah i dont know how to use the polaris CLI, so i just copied directly from https://github.com/apache/polaris/blob/main/regtests/run_spark_sql.sh

kevinjqliu commented on getting-started/spark/README.md:
yes I'll admit, this README is a filler for now as a way to get spark & polaris up and running quickly

kevinjqliu commented on getting-started/spark/README.md:
```diff
@@ -0,0 +1,51 @@
+* The `create-polaris-catalog` service to run setup script and create local catalog in Polaris
+
+## Access the Jupyter notebook interface
+In the Jupyter notebook container log, look for the URL to access Jupyter notebook. The url should be in the format, `http://127.0.0.1:8888/lab?token=<token>`.
```
which part is the typo?

sfc-gh-emaynard replied:
The existing text just says `to access Jupyter notebook`

kevinjqliu replied:
changed to `to access the Jupyter notebook`

kevinjqliu commented:
md check intermittently shows `https://redocly.com/docs/cli/installation` as 400 error, weird

flyrain commented:
> md check intermittently shows https://redocly.com/docs/cli/installation as 400 error, weird

It's OK to remove the link for now since we’re transitioning to Hugo.

kevinjqliu commented:
@flyrain just had to run the CI a few times, it's unrelated to this change

kevinjqliu commented on .github/workflows/check-md-link.yml:
```diff
@@ -41,5 +41,5 @@ jobs:
       with:
         use-quiet-mode: 'yes'
         config-file: '.github/workflows/check-md-link-config.json'
-        folder-path: 'regtests, regtests/client/python/docs, regtests/client/python, .github, build-logic, polaris-core, polaris-service, extension, spec, k8, notebooks'
```
This PR moved `notebooks/` from top-level directory into the `getting-started/` directory

flyrain commented on getting-started/spark/create-polaris-catalog.sh:
```diff
@@ -0,0 +1,56 @@
+SPARK_BEARER_TOKEN="${REGTEST_ROOT_BEARER_TOKEN:-principal:root;realm:default-realm}"
+POLARIS_CATALOG_NAME="${POLARIS_CATALOG_NAME:-polaris_demo}"
+
+# create a catalog backed by the local filesystem
```
I'm not entirely sure if we need this file. Could we handle everything directly within the notebook, like the other operations in `SparkPolaris.ipynb`? Would it simplify things if we moved the operations there?

kevinjqliu replied:
we could, but i think its a good idea to separate infra code (this script) from application code (the notebook)

flyrain replied:
We could initialize catalog in notebook as well. I feel it's more flexible that way. for example, you don't have to worry about the an env variable for catalog name. But I'm OK with either one. Not a blocker for me.

flyrain commented on getting-started/spark/README.md:
```diff
@@ -0,0 +1,51 @@
+
+# Getting Started with Apache Spark and Apache Polaris
+
+This getting started guide provides a `docker-compose` file to set up [Apache Spark](https://spark.apache.org/) with Apache Polaris. Apache Polaris is configured as an Iceberg REST Catalog in Spark. 
```
There is other way to try Spark with Polaris without docker, it's a not a blocker, we can add it later.

flyrain reviewed (commented):
Thanks @kevinjqliu for working on it. LGTM overall. Left some comments and questions.

kevinjqliu commented:
Thanks for the review @flyrain, addressed your comments

flyrain commented:
We cannot merge any PR until #374 is merged.

kevinjqliu commented:
Thanks for the heads up, I'll rebase once that PR's merged

kevinjqliu commented:
@flyrain took your advice, moved `getting-started/spark/create-polaris-catalog.sh` logic into the jupyter notebook. 
Also rebased off latest main. I think this PR is good to go. Please take a look!

flyrain commented:
Thanks a lot for working on it, @kevinjqliu! Thanks all for the review.
